src/coreclr/jit/helperexpansion.cpp

```diff
@@ -0,0 +1,1043 @@
+//
+// Returns:
+//    true if a helper was expanded
+//
+template <bool (Compiler::*ExpansionFunction)(BasicBlock*, Statement*, GenTreeCall*)>
+bool Compiler::fgExpandHelperForBlock(BasicBlock* block)
+{
+    for (Statement* const stmt : block->NonPhiStatements())
+    {
+        if ((stmt->GetRootNode()->gtFlags & GTF_CALL) == 0)
+        {
+            // TP: Stmt has no calls - bail out
+            continue;
+        }
+
+        for (GenTree* const tree : stmt->TreeList())
+        {
+            if (!tree->IsHelperCall())
+            {
+                continue;
+            }
+
+            if ((this->*ExpansionFunction)(block, stmt, tree->AsCall()))
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// fgExpandStaticInit: Partially expand static initialization calls, e.g.:
+//
+//    tmp = CORINFO_HELP_X_NONGCSTATIC_BASE();
+//
+// into:
+//
+//    if (isClassAlreadyInited)
+//        CORINFO_HELP_X_NONGCSTATIC_BASE();
+//    tmp = fastPath;
+//
+// Returns:
+//    PhaseStatus indicating what, if anything, was changed.
+//
+PhaseStatus Compiler::fgExpandStaticInit()
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHaveStaticInit())
+    {
+        // TP: nothing to expand in the current method
+        JITDUMP("Nothing to expand.\n")
+        return result;
+    }
+
+    if (opts.OptimizationDisabled())
+    {
+        JITDUMP("Optimizations aren't allowed - bail out.\n")
+        return result;
+    }
+
+    // TODO: Replace with opts.compCodeOpt once it's fixed
+    const bool preferSize = opts.jitFlags->IsSet(JitFlags::JIT_FLAG_SIZE_OPT);
+    if (preferSize)
+    {
+        // The optimization comes with a codegen size increase
+        JITDUMP("Optimized for size - bail out.\n")
+        return result;
+    }
+
+    return fgExpandHelper<&Compiler::fgExpandStaticInitForCall>(true);
+}
+
+//------------------------------------------------------------------------------
+// fgExpandStaticInitForCall: Partially expand given static initialization call.
+//    Also, see fgExpandStaticInit's comments.
+//
+// Arguments:
+//    block  - call's block
+//    stmt   - call's statement
+//    call   - call that represents a static initialization
+//
+// Returns:
+//    true if a static initialization was expanded
+//
+bool Compiler::fgExpandStaticInitForCall(BasicBlock* block, Statement* stmt, GenTreeCall* call)
+{
+    assert(call->IsHelperCall());
+
+    bool                    isGc       = false;
+    StaticHelperReturnValue retValKind = {};
+    if (!IsStaticHelperEligibleForExpansion(call, &isGc, &retValKind))
+    {
+        return false;
+    }
+
+    assert(!call->IsTailCall());
+
+    if (call->gtInitClsHnd == NO_CLASS_HANDLE)
+    {
+        assert(!"helper call was created without gtInitClsHnd or already visited");
```

Comment:
Just curious - so how exactly did you fix the problem we discussed?

Reply:
Basically this problem was because I was wrongly passing parameter which was not skipping rarely run blocks. I am just relying on the existing logic of that. If we want more sophisticated ways like returning the appropriate `bbNext`, that should be a follow-up because compact blocks might change things and sometimes we might not return valid block.